src/lua/uri.c

```diff
@@ -0,0 +1,313 @@
+{
+	bool nested_tables_is_present = false;
+	int uri_count = lua_objlen(L, -1);
+	for (int i = 0; i < uri_count && !nested_tables_is_present; i++) {
+		lua_rawgeti(L, -1, i + 1);
+		nested_tables_is_present = (lua_type(L, -1) == LUA_TTABLE);
+		lua_pop(L, 1);
+	}
+	return nested_tables_is_present;
+}
+
+/**
+ * Perform some checks for the URI table. The types of checks depend on
+ * @a table_is_nested.
+ */
+static bool
+uri_table_is_valid(struct lua_State *L, bool table_is_nested)
+{
+	/** All numerical keys in URI table corresponds to URI */
+	int uri_count_from_numerical_keys = lua_objlen(L, -1);
+	bool params_is_present = is_field_present(L, "params");
+	bool uri_key_is_present = is_field_present(L, "uri");
+	bool nested_uri_tables_is_present = is_nested_uri_tables_present(L);
+	bool default_params_is_present = is_field_present(L, "default_params");
+
+        /* Should be at least one URI in URI table */
+	if (uri_count_from_numerical_keys == 0 && !uri_key_is_present)
+		return false;
+	/* Mix "uri=" syntax and URIs from numerical keys is banned */
+	if (uri_count_from_numerical_keys != 0 && uri_key_is_present)
+		return false;
+	/* Should be no more than one URI in nested URI table */
+	if (uri_count_from_numerical_keys > 1 && table_is_nested)
+		return false;
+	/* Table "params" is now allowed for several URIs */
+	if (uri_count_from_numerical_keys > 1 && params_is_present)
+		return false;
+	/* Nested tables is not allowed in nested tables */
+	if (table_is_nested && nested_uri_tables_is_present)
+		return false;
+	/* Table "default_params" is not allowed in nested tables */
+	if (default_params_is_present && table_is_nested)
+		return false;
+	/*
+	 * Table "params" is not allowed in table, which contains,
+	 * nested URI tables.
+	 */
+	if (params_is_present && nested_uri_tables_is_present)
+		return false;
+	return true;
+}
+
+/**
+ * Create @a uri_set from the table, which located at the top of lua stack.
+ * Supports many different formats.
+ */
+static int
+uri_set_create_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			bool table_is_nested)
+{
+	int rc = 0;
+	struct uri_set u = { 0 };
+
+	/** Some preliminary checks of the URI table. */
+	if (!uri_table_is_valid(L, table_is_nested))
+		return -1;
+
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
+		int key_type = lua_type(L, -2);
+		int val_type = lua_type(L, -1);
+		const char *key;
+		const char *val;
+		switch (key_type) {
+		case LUA_TNUMBER:
+			switch (val_type) {
+			case LUA_TSTRING:
+			case LUA_TNUMBER:
+				val = lua_tostring(L, -1);
+				rc = uri_set_add_uri(&u, val);
+				break;
+			case LUA_TTABLE:
+				rc = uri_set_create_from_lua(&u, L, true);
+				break;
+			default:
+				rc = -1;
+			};
+			break;
+		case LUA_TSTRING:
+			key = lua_tostring(L, -2);
+			if (strcmp(key, "uri") == 0) {
+				switch (val_type) {
+				case LUA_TSTRING:
+				case LUA_TNUMBER:
+					val = lua_tostring(L, -1);
+					rc = uri_set_add_uri(&u, val);
```

Comment:
Let's try to reorganize the code a little bit:

1. Add a function that creates a `uri` object from a Lua string or table.
2. Make `uri_set_add` take a `uri` object instead of a string. It doesn't need to copy the `uri` - it may take ownership. We will use this function to populate a `uri_set` with uris created from Lua objects with function 1.
3. In this function, let's first figure out what we have at hand - array of uris or a single uri. If we believe it to be a single uri, we will use function 1 to create a `uri` object, then add it to an empty `uri_set`. If it's an array or uris, then walk over the array, and call function 1 for each entry. Then apply `default_params` to all uris.

We can figure out if it's a single uri or an array of uris as follows:
- if the table length is 0 and `uri` key is present, then it's a single uri;
- if the table length is 1 and the type of the first entry is a string, then it's a single uri.
- otherwise it must be an array of uris.

All other sanity checks related to a single uri should live in function 1.

Reply:
I rework patchset, maybe it is not how you write here, but i think it's good now. Can you see it again and if it's bad disscuss about it again